### DISCUSSION TITLE
Add Click-to-Publish Goal Feature for Protobuf Messages in 3D Panel

### DISCUSSION THREAD

**Discussion by DYwasd** [foxglove/community#1014] [General]
#### Background and Motivation
Currently, the 3D panel in Foxglove Studio supports the click-to-publish goal feature for ROS message formats. This feature greatly enhances usability by allowing users to easily set goals within the 3D environment. However, this functionality is not available for Protobuf message formats. Adding this feature would provide a consistent user experience and extend the usability of the 3D panel for those using Protobuf messages.

#### Feature Description
- Enable the click-to-publish goal feature for Protobuf message formats in the 3D panel.
- The implementation should mirror the existing functionality available for ROS message formats.
- Ensure that users can configure and customize the goals for Protobuf messages just as they can for ROS messages.

#### Additional Information
Providing this feature will enhance the flexibility and functionality of the 3D panel, making Foxglove Studio more versatile and user-friendly for a wider range of users.

Thank you for considering this feature request.

**Comment by jtbandes** [comment]
Internal tracking issue: [FG-8253](https://linear.app/foxglove/issue/FG-8253/click-to-publish-3d-panel-non-ros-support)

**Reply by lucafei** [reply]
when will this feature be added into foxglove?

**Comment by lucafei** [comment]
yes, really need this feature, hope it will be integreated in next updating soon

**Comment by czh9526** [comment]
Are you referring to features like (Publish 2D pose estimate (/initialpose))? I am waiting for when this feature will be updated

**Comment by janblumenkamp** [comment]
I need this feature too, but found a workaround if you are using the [foxglove websocket](https://github.com/foxglove/ws-protocol) in Python. You can manually specify the capabilities and metadata like so:
```
FoxgloveServer(
    "0.0.0.0", 8765, "zenoh-foxglove bridge",
    capabilities=["clientPublish"],
    metadata={"ROS_DISTRO": "humble"}
)
```
The message publish utility will then become visible in the 3D panel and the `FoxgloveServerListener` will then trigger the `on_client_message` callback when the message is published.

**Reply by lucafei** [reply]
what's data type of sent message through click to publish, i need to decode this sent message by click to publish on non-ros distro.

**Reply by defunctzombie** [reply]
Hi - we recently extended this feature to not require metadata and work with systems other than ROS. We've documented the message you'll receive in those cases here: https://docs.foxglove.dev/docs/visualization/panels/3d#publish

What your server receives is a JSON object with the fields described in the respective schemas (ROS or Foxglove). Your server will need to process that JSON to whatever format is appropriate for your middlware or system.

**Reply by lucafei** [reply]
hi，developer. I found that the messages published via the "click to publish" button on Foxglove are not strongly bound to topics, which means that when I receive messages in the server through serverOptions.callbacks.onMessageData, I cannot distinguish messages for a specific topic. Are there any methods to bind my callback function to a specific topic now, so that I can only receive the messages I want—just like the publish-subscribe mechanism in ROS? Or will you address this issue in subsequent updates?

**Reply by gasmith** [reply]
@lucafei the [`onMessageData` callback](https://foxglove-sdk-api-docs.pages.dev/cpp/generated/api/structfoxglove_1_1WebSocketServerCallbacks#_CPPv4N8foxglove24WebSocketServerCallbacks13onMessageDataE) includes a `client_channel_id`. If you also define an [`onClientAdvertise` callback](https://foxglove-sdk-api-docs.pages.dev/cpp/generated/api/structfoxglove_1_1WebSocketServerCallbacks#_CPPv4N8foxglove24WebSocketServerCallbacks17onClientAdvertiseE), you'll receive the topic (and schema) corresponding to that channel ID. There's an example implementation of the `onClientAdvertise` callback [here](https://github.com/foxglove/foxglove-sdk/blob/dbfe19ff95e8a0e9da8b47cc96710803eb3fc848/cpp/examples/ws-server/src/main.cpp#L41-L55). You can't define a callback for a specific topic, however; you'll need to do your own filtering & routing.
